src/main/java/mate.academy/lib/Injector.java

```diff
@@ -1,13 +1,93 @@
+import mate.academy.service.impl.ProductParserImpl;
+import mate.academy.service.impl.ProductServiceImpl;
+
 public class Injector {
     private static final Injector injector = new Injector();
+    private final Map<Class<?>, Object> instances = new HashMap<>();
 
     public static Injector getInjector() {
         return injector;
     }
 
-    public Object getInstance(Class<?> interfaceClazz) {
-        return null;
+    public Object getInstance(Class<?> clazz) {
+        if (instances.containsKey(clazz)) {
+            return instances.get(clazz);
+        }
+
+        try {
+            // Check if the class is an interface
+            if (clazz.isInterface()) {
+                Class<?> implementationClass = findImplementationClass(clazz);
+                Object instance = createInstance(implementationClass);
+                injectFields(instance);
+                instances.put(clazz, instance);
+                return instance;
+            } else {
+                Object instance = createInstance(clazz);
+                injectFields(instance);
+                instances.put(clazz, instance);
+                return instance;
+            }
+        } catch (Exception e) {
+            throw new RuntimeException("Failed to create an instance of class: "
+                    + clazz.getName(), e);
+        }
+    }
+
+    private Class<?> findImplementationClass(Class<?> interfaceClass) {
+        // Implement your logic to find the implementation class for the interface
+        // This is just a placeholder; you should customize it based on your application's
+        if (interfaceClass.equals(FileReaderService.class)) {
+            return FileReaderServiceImpl.class;
+        } else if (interfaceClass.equals(ProductParser.class)) {
+            return ProductParserImpl.class;
+        } else if (interfaceClass.equals(ProductService.class)) {
+            return ProductServiceImpl.class;
+        }
+        throw new RuntimeException("No implementation class found for: "
+                + interfaceClass.getName());
+    }
+
+    private Object createInstance(Class<?> clazz) throws Exception {
+        // Find the constructor with the most parameters
+        Constructor<?> constructor = null;
+        for (Constructor<?> c : clazz.getDeclaredConstructors()) {
+            if (constructor == null || c.getParameterCount() > constructor.getParameterCount()) {
+                constructor = c;
+            }
+        }
+        if (constructor == null) {
+            throw new RuntimeException("No suitable constructor found for class: "
+                    + clazz.getName());
+        }
+
+        // Create an instance using the constructor
+        Object[] params = new Object[constructor.getParameterCount()];
+        for (int i = 0; i < constructor.getParameterCount(); i++) {
+            params[i] = getInstance(constructor.getParameterTypes()[i]);
+        }
+        return constructor.newInstance(params);
```

Comment:
no check for exception, no saving to the instances Map, too much redundant logic..
```suggestion
        try {
            Constructor<?> constructor = clazz.getConstructor();
            Object instance = constructor.newInstance();
            instances.put(clazz, instance);
            return instance;
        } catch (ReflectiveOperationException e) {
            throw new RuntimeException("Can't create a new instance of " + clazz.getName());
        }
```